doc/developer/design/20230829_topk_size_hint.md

```diff
@@ -0,0 +1,426 @@
+# Top-K Group Size Query Hints
+
+- Associated: MaterializeInc/materialize#18883
+
+## The Problem
+
+Min/max reductions and top-k operations are computed in Materialize by creating a stack of
+arrangements and performing tournament-tree-style hierarchical processing of updates. By default,
+the height of the hierarchies created is sized to accommodate the existence of potentially
+very large query groups in the data. This implies that Materialize provides an incremental update
+experience from the start, even in the presence of large input query groups, but this comes at the
+cost of extra memory usage.
+
+Memory usage can be reduced substantially for queries with min/max or top-k by tuning the
+`EXPECTED GROUP SIZE` query hint. The query hint informs Materialize about the maximum size of
+the query groups that are expected in the input to the min/max reduction or top-k operation. Based
+on this information, Materialize can size down the height of the hierarchy and allocate stack fewer
+arrangements.
+
+Currently, the `EXPECTED GROUP SIZE` controls the tuning of both min/max reductions and top-k computations.
+However, when the two patterns co-occur in the same query block, it is currently impossible to tune them
+independently. This problem is reported in issue MaterializeInc/materialize#18883. We expand the first
+example query provided in the issue to illustrate the problem in more detail:
+
+```sql
+SELECT
+    sum(max_revenue) AS sum_max_revenue
+FROM (
+    SELECT l_orderkey,
+           max(l_extendedprice * (1 - l_discount)) AS max_revenue
+    FROM lineitem
+    GROUP BY l_orderkey
+    OPTIONS (EXPECTED GROUP SIZE = 7)
+    ORDER BY max_revenue DESC
+    LIMIT 10
+);
+```
+
+The `EXPLAIN` output for the above query is:
+
+```
+                                   Optimized Plan
+-------------------------------------------------------------------------------------
```

Comment:
nit: I think it's more instructive if you print the `RAW PLAN` outputs in this doc.

Reply:
The issue I had with `RAW PLAN` is that it misses the `exp_group_size` at the root level of a query because it displays a `Finishing` and not a `TopK`. And if I tried `EXPLAIN RAW PLAN FOR MATERIALIZED VIEW ...`, then I got an error. But it was useful to contrast the `RAW PLAN` with the `OPTIMIZED PLAN` to detect bug #21501.

Reply:
Got it.